### PR TITLE
Feature/processing framework

### LINKED ISSUE #256: Support Process framework in OSN

### PR DESCRIPTION
**Pull request by LetThereBeDwight**
Includes changes to utilize the processing framework in OSN.
Also rolls in updates for WFS/legacyID extraction, and makes #244 OBE.

Need to test WFS and MapServiceImages, but this should all be ready for review otherwise.
The progress displays...leave something to be desired.

resolve #256 

resolve DigitalGlobe/GGD-DeepCore-Dev#321
resolve DigitalGlobe/GGD-DeepCore-Dev#328
resolve DigitalGlobe/GGD-DeepCore-Dev#262

### CODE REVIEW

File: common/src/OpenSpaceNet.cpp [hunk 1]
```diff
@@ -102,109 +87,182 @@ OpenSpaceNet::OpenSpaceNet(OpenSpaceNetArgs&& args) :
...
+    startProgressDisplay();
+    auto startTime = high_resolution_clock::now();
 
-    if (!args_.resampledSize) {
-        for (auto c : args_.windowSize) {
-            DG_CHECK(c <= model_->metadata().modelSize().width,
-                     "Argument --window-size contains a size that does not fit within the model (width: %d).",
-                      model_->metadata().modelSize().width)
-        }
+    if (!args_.quiet && pd_) {
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: You don't need to create a separate thread for each metric, also, detaching threads like this is asking for trouble.

This is what you do:
```cpp
if (!args_.quiet && pd_) {
        Semaphore stop;
        auto allMetrics = async([&stop, &slidingWindow, &featureSink, ...... ] () {
            do {
                if(slidingWindow->metric("processed").changed().tryWait()) {
                    auto processed = slidingWindow->metric("processed").convert<int>();
                    auto requested = slidingWindow->metric("requested").convert<int>();
                    if (processed > 0 && requested > 0) {
                        auto progress = (float) processed / requested;
                        this->pd_->update("Reading", progress);
                    }
                }
                ......

                stop.wait([&slidingWindow, &featureSink, ...... ] () {
                    return featureSink->state() == NodeState::STOPPED ||
                           slidingWindow->metric("processed").changed().check() || 
                           featureSink->metric("processed").changed().check() ||
                           ...... ;
                });

            } while(featureSink->state() != NodeState::STOPPED);
        });

        featureSink->run();
        featureSink->wait();
        stop.notify();
        allMetrics.get();
        pd_->stop();
        ......
```

File: common/src/OpenSpaceNet.cpp [hunk 2]
```diff
@@ -238,102 +296,81 @@ void OpenSpaceNet::initLocalImage()
...
     DG_CHECK(args_.bbox, "Bounding box must be specified");
 
+    std::unique_ptr<deepcore::imagery::MapServiceClient> client;
     bool wmts = true;
     string url;
     switch(args_.source) {
         case Source::MAPS_API:
-            OSN_LOG(info) << "Connecting to MapsAPI..." ;
-            client_ = make_unique<MapBoxClient>(args_.mapId, args_.token);
+            OSN_LOG(info) << "    Connecting to MapsAPI..." ;
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: Why did you add spaces here?

File: common/src/OpenSpaceNet.cpp [hunk 3]
```diff
@@ -238,102 +296,81 @@ void OpenSpaceNet::initLocalImage()
...
+        client->setLayer("DigitalGlobe:ImageryTileService");
+        client->setTileMatrixSet("EPSG:3857");
+        client->setTileMatrixId((format("EPSG:3857:%1d") % args_.zoom).str());
     } else {
-        client_->setTileMatrixId(lexical_cast<string>(args_.zoom));
+        client->setTileMatrixId(lexical_cast<string>(args_.zoom));
     }
 
-    unique_ptr<Transformation> llToProj(client_->spatialReference().fromLatLon());
+    unique_ptr<Transformation> llToProj(client->spatialReference().fromLatLon());
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]:
```cpp
auto llTopProj = client->spatialReference().fromLatLon();
```

File: common/src/OpenSpaceNet.cpp [hunk 4]
```diff
@@ -238,102 +296,81 @@ void OpenSpaceNet::initLocalImage()
...
+SubsetRegionFilter::Ptr OpenSpaceNet::initSubsetRegionFilter()
 {
     if (args_.filterDefinition.size()) {
-        OSN_LOG(info) << "Initializing the region filter..." ;
+        OSN_LOG(info) << "Initializing the subset filter..." ;
 
-        regionFilter_ = make_unique<MaskedRegionFilter>(cv::Rect(0, 0, bbox_.width, bbox_.height),
-                                                        calcPrimaryWindowStep(),
-                                                        MaskedRegionFilter::FilterMethod::ANY);
+        RegionFilter::Ptr regionFilter = make_unique<MaskedRegionFilter>(cv::Rect(0, 0, bbox_.width, bbox_.height),
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: `RegionFilter` is an `Object`, initialize it as such:

```cpp
RegionFilter::Ptr regionFilter = MaskedRegionFilter::create(cv::Rect(0, 0, bbox_.width, bbox_.height),
                                                                         calcPrimaryWindowStep(),
                                                                         MaskedRegionFilter::FilterMethod::ANY);
```

File: common/src/OpenSpaceNet.cpp [hunk 5]
```diff
@@ -363,359 +400,282 @@ void OpenSpaceNet::initFilter()
...
-                if(!blockQueue.empty()) {
-                    item = blockQueue.back();
-                    blockQueue.pop_back();
-                } else {
-                    queueMutex.unlock();
-                    haveWork.wait();
-                    continue;
-                }
+    float confidence = 0;
+    if(args_.action == Action::LANDCOVER && model->modelOutput() == ModelOutput::BOX) {
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: Remove references to `LANDCOVER`.

File: common/src/OpenSpaceNet.cpp [hunk 6]
```diff
@@ -363,359 +400,282 @@ void OpenSpaceNet::initFilter()
...
+void OpenSpaceNet::initSegmentation(Model::Ptr model)
 {
-    auto predictions = processSerial<PredictionPoly>();
-
-    if(args_.action != Action::LANDCOVER && args_.nms) {
-        skipLine();
-        OSN_LOG(info) << "Performing non-maximum suppression..." ;
-        predictions = polygonalNonMaxSuppression(predictions, args_.overlap / 100);
-    }
+    auto castModel = dynamic_cast<CaffeSegmentation*>(model.get());
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: No reason that we can't support other segmentation models. Also use `std::dynamic_pointer_cast`
```cpp
auto segmentation = dynamic_pointer_cast<Segmentation>(model.get());
```

File: common/src/OpenSpaceNet.cpp (hunk 2 above)
Comment (LetThereBeDwight) [reply common/src/OpenSpaceNet.cpp]: I had messed around with how some info was output but ended up with much of it exactly the same - reverting this..

File: common/src/OpenSpaceNet.cpp (hunk 1 above)
Comment (avitebskiy) [reply common/src/OpenSpaceNet.cpp]: This actually brings up a need to do something better about monitoring `Metric`s. I think the work I'll need to do fix the data race in `Node` will enable something easier to use, but the solution I posted here should suffice for now...

File: common/src/OpenSpaceNet.cpp [hunk 7]
```diff
@@ -363,359 +400,282 @@ void OpenSpaceNet::initFilter()
...
-                    continue;
-                }
+    float confidence = 0;
+    if(args_.action == Action::LANDCOVER && model->modelOutput() == ModelOutput::BOX) {
+        auto windowSize = calcPrimaryWindowSize();
+        if (args_.windowSize.size() < 2) {
+            if(blockSize_.width % windowSize.width == 0 &&
+               blockSize_.height % windowSize.height == 0) {
+                bbox_ = {cv::Point {0, 0}, imageSize_};
+                concurrent_ = true;
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: None of the "concurrent" logic is necessary. Remove all references.

File: common/src/OpenSpaceNet.cpp [hunk 8]
```diff
@@ -363,359 +400,282 @@ void OpenSpaceNet::initFilter()
...
-        bool skip = false;
-        for(const auto& point : prediction.poly.exteriorRing.points) {
-            if(point.x < bbox_.x || point.y < bbox_.y || point.x > bbox_.width || point.y > bbox_.height) {
-                skip = true;
-                break;
-            }
-        }
+dg::deepcore::imagery::node::SlidingWindow::Ptr OpenSpaceNet::initSlidingWindow()
+{
+    auto slidingWindow = dg::deepcore::imagery::node::SlidingWindow::create("SlidingWindow");
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: Didn't you import the whole `dg::deepcore::imagery::node` namespace? If you did, there shouldn't be a need to fully qualify it here. Or is there ambiguity because whole namespaces are imported?

File: common/src/OpenSpaceNet.cpp [hunk 9]
```diff
@@ -363,359 +400,282 @@ void OpenSpaceNet::initFilter()
...
+    return predictionToFeature;
+}
 
-    auto startTime = high_resolution_clock::now();
-    auto mat = GeoImage::readImage(*image_, bbox_, regionFilter_.get(), [this](float progress) -> bool {
-        if(pd_ && !args_.quiet) {
-            pd_->update("Reading", progress);
+WfsFeatureFieldExtractor::Ptr OpenSpaceNet::initWfs()
+{
+    if (args_.dgcsCatalogID || args_.evwhsCatalogID) {
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: All this stuff should probably be moved to DeepCore.

File: common/src/OpenSpaceNet.cpp [hunk 10]
```diff
@@ -363,359 +400,282 @@ void OpenSpaceNet::initFilter()
...
-            { "top_cat", { FieldType::STRING, predictions[0].label.c_str() } },
-            { "top_score", { FieldType ::REAL, predictions[0].confidence } },
-            { "date", { FieldType ::DATE, time(nullptr) } }
-    };
+    ProgressCategories categories = {{"Reading", "Reading the image" }};
 
-    ptree top5;
-    for(const auto& prediction : predictions) {
-        top5.put(prediction.label, prediction.confidence);
+    if (args_.action == Action::LANDCOVER) {
```
Comment (avitebskiy) [inline common/src/OpenSpaceNet.cpp]: Remove all references to `LANDCOVER`.

File: common/src/OpenSpaceNet.cpp (hunk 9 above)
Comment (geovations) [reply common/src/OpenSpaceNet.cpp]: concur

File: common/src/OpenSpaceNet.cpp (hunk 8 above)
Comment (LetThereBeDwight) [reply common/src/OpenSpaceNet.cpp]: For nodes that share names with their non-node counterparts, there is ambiguity. There is a similar issue with the detector nodes.

File: common/src/OpenSpaceNet.cpp (hunk 9 above)
Comment (LetThereBeDwight) [reply common/src/OpenSpaceNet.cpp]: Do you mean that all of the stuff that goes into initializing the WFS should go into deepcore? Currently the WFS stuff is highly generalized, is the thought that there should be subclasses to those nodes like "DGCSFeatureFieldExtractor" etc?
